Tasks 2 - 8 Savanna Web

22DPVEmer opened the pull request:
Completed tasks 2 - 8 in this pull request. Created and updated display of the game page, and buttons for all four animal types, 2 of which are plugins. Options to change from emoji to character display, pause game, quit game, resume, and start actions were added to the web app, and new tables to save and load games including AnimalStates for each animal, GameStates, and GameSaves.
1. Image of new game page:
![image](https://github.com/user-attachments/assets/54a278fd-9671-4682-a763-eea3b26dbed8)
2. Display of emojis
![image](https://github.com/user-attachments/assets/bf7c79fd-10a9-40b3-9c25-1f8e3a3a9fc6)
3. Display of letters
![image](https://github.com/user-attachments/assets/389a60f0-2fff-4219-b2b7-14fa058f4129)
4.Save game menu with animal count of type antelope and lion, iteration and date:
![image](https://github.com/user-attachments/assets/ac05abfc-613c-4db0-846d-519a44d45413)
5.Image of successful load from saved games:
![image](https://github.com/user-attachments/assets/7c803808-f163-4938-ab76-58452ae23914)
6. Message that progress will not be saved by closing or opening other tab:
![image](https://github.com/user-attachments/assets/c641fb2d-861b-495d-a5cc-f87da3f78c32)
7. Message to confirm delete of a save:
![image](https://github.com/user-attachments/assets/0839010a-b58f-48ae-93c4-f243ed69162a)

22DPVEmer committed:
Enhance project structure and add Savanna.Services project

- Added Savanna.Services project to solution
- Updated Savanna.Web to integrate new services
- Configured plugin and game service initialization in Program.cs
- Added logging and dependency injection for game services
- Updated project references and build configurations
- Improved game initialization and service management

22DPVEmer committed:
Refactor client-side game management with centralized constants and error handling

- Replaced hardcoded console logs and error messages with GameConstants
- Simplified error handling in AnimalManager, GameControls, GameState, and SaveManager
- Removed verbose logging and error tracking
- Improved code readability and consistency across client-side game scripts

22DPVEmer committed:
Improve game state tracking and authentication handling

- Update GamesController to count only living animals in game saves
- Enhance user authentication in save retrieval by using actual user ID
- Modify Dashboard page to use Authorize attribute
- Remove Dashboard code-behind file
- Update SaveManager and UIManager to handle living animals more accurately
- Improve error handling in save deletion process

sreeharihm commented on Savanna.Infrastructure/DesignTimeDbContextFactory.cs:
```diff
@@ -12,7 +12,7 @@ public class DesignTimeDbContextFactory : IDesignTimeDbContextFactory<Applicatio
         {
             IConfigurationRoot configuration = new ConfigurationBuilder()
-                .SetBasePath(Directory.GetCurrentDirectory())
+                .SetBasePath(Path.Combine(Directory.GetCurrentDirectory(), "../Savanna.Web"))
```
Move this to a constant class

22DPVEmer replied:
Moved it

sreeharihm commented on Savanna.Services/Services/AnimalConfigurationService.cs:
```diff
@@ -0,0 +1,131 @@
+
+                if (config.Plugins == null)
+                {
+                    _logger.LogError("Plugins section is null in config.json");
```
try to add a middleware for handling the exceptions

22DPVEmer replied:
Added middleware for most of the exceptions in the service layer.

sreeharihm commented on Savanna.Web/Controllers/GamesController.cs:
```diff
@@ -0,0 +1,329 @@
+                        Iteration = g.GameState.CurrentIteration,
+                        AnimalCounts = new
+                        {
+                            Lion = g.GameState.Animals.Count(a => a.AnimalType == "Lion" && a.IsAlive),
```
move these hard coding to a constant class

22DPVEmer replied:
Moved it

sreeharihm commented on Savanna.Web/wwwroot/js/game/animalManager.js:
Similar to Game Constants, you can move all the endpoints to EndpointConstant

22DPVEmer replied:
Moved it

sreeharihm commented on Savanna.Web/wwwroot/js/game/gameState.js:
```diff
@@ -0,0 +1,68 @@
+        try {
+            const response = await fetch('/api/games/update', { method: 'POST' });
+            if (!response.ok) {
+                if (response.status === 401) {
```
move statuscode to a constant

22DPVEmer replied:
Moved that also to a constant.

sreeharihm reviewed (commented):
Added review comments

22DPVEmer committed:
Refactor client-side game management with improved error handling and constants

- Update DesignTimeDbContextFactory to use project path constants
- Enhance AnimalConfigurationService with custom configuration exceptions
- Modify GameService to use custom service exceptions
- Improve client-side JavaScript with centralized game constants
- Refactor UIManager and SaveManager for better error handling and message display
- Add more comprehensive animal type and display mode support

sreeharihm commented on Savanna.Services/Services/AnimalConfigurationService.cs:
move these class to separate model class

22DPVEmer replied:
I moved each to its file under Models namespace.

sreeharihm commented on Savanna.Web/Middleware/ExceptionHandlingMiddleware.cs:
```diff
@@ -0,0 +1,102 @@
+
+                default:
+                    context.Response.StatusCode = (int)HttpStatusCode.InternalServerError;
+                    response.Message = "An error occurred while processing your request";
```
move these messages to constant

22DPVEmer replied:
I moved the messages of middleware to a constant class.

sreeharihm reviewed (commented):
added review comments

22DPVEmer committed:
Refactor configuration and exception handling with centralized constants

- Move config file path to ProjectPaths constant
- Extract configuration models from AnimalConfigurationService to separate Models
- Update ExceptionHandlingMiddleware to use MiddlewareConstants for status codes and messages
- Improve code consistency by centralizing configuration and error handling constants

sreeharihm reviewed (approved):
Looks good!!
